helix-term/src/ui/markdown.rs

```diff
@@ -70,17 +79,19 @@ impl Markdown {
             })
         }
 
-        macro_rules! get_theme {
-            ($s1: expr) => {
-                theme
-                    .map(|theme| theme.try_get($s1.as_str()))
-                    .flatten()
-                    .unwrap_or_default()
-            };
-        }
-        let text_style = get_theme!(self.text_style);
-        let code_style = get_theme!(self.block_style);
-        let heading_style = get_theme!(self.heading_style);
+        let get_theme = |key: &str| {
+            theme
+                .map(|theme| theme.try_get(key))
+                .flatten()
+                .unwrap_or_default()
+        };
+        let text_style = get_theme(&self.text_style);
+        let code_style = get_theme(&self.block_style);
+        let heading_styles: Vec<Style> = self
+            .heading_styles
+            .iter()
+            .map(|key| get_theme(key))
```

Comment:
This has no fallback though, if `markup.heading.1` isn't defined, just `markup.heading` it won't highlight.

Reply:
The macro would have to be modified to try `get_theme(key, "markup.heading")` in order. We can also simplify and just use `markup.heading` for this component.

Reply:
Not sure if recursion is a good idea but something along those lines should work:

```rust
// example program

use std::collections::HashMap;

fn get_theme<'a, Y>(theme: &'a HashMap<&str, Y>, key: &str) -> Option<&'a Y> {
    theme.get(key).or_else(|| match key.rfind(".") {
        Some(n) => get_or_fallback(theme, &key[..n]),
        None => None,
    })
}

fn main() {
    let mut theme = HashMap::new();

    theme.insert("markup.heading", 1);
    theme.insert("markup.heading.2", 2);

    assert_eq!(get_theme(&theme, "markup.heading.1"), Some(&1));
    assert_eq!(get_theme(&theme, "markup.heading.2"), Some(&2));
    assert_eq!(get_theme(&theme, "markup.bold"), None);
}
```

Alternatively, the fallbacks for every key (`Vec<String>`) could be specified manually or generated in the constructor, if this is a rather hot function.

I'm not very familiar with macros, sorry. Do they have a performance/convinience advantage in this case?

Reply:
Yeah I don't think there was any reason for macros, personally in not sure if we really need the namespacing for different popups https://github.com/helix-editor/helix/pull/1363

Here's a similar method in syntax.rs, perhaps we can use the same algorithm: https://github.com/helix-editor/helix/blob/d3221b03a2b3ee5acc68066f34c4e56e2d05ce94/helix-core/src/syntax.rs#L1197-L1211

Reply:
> Yeah I don't think there was any reason for macros, personally in not sure if we really need the namespacing for different popups.

I would be in favour of removing the namespacing part, it seems unnecessary and too granular to be widely used in themes.